packages/flutter/lib/src/rendering/sliver_group.dart

```diff
@@ -327,12 +327,50 @@ class RenderSliverMainAxisGroup extends RenderSliver with ContainerRenderObjectM
 
   @override
   void paint(PaintingContext context, Offset offset) {
-    RenderSliver? child = lastChild;
+    if (firstChild == null) {
+      return;
+    }
+    // offset is to the top-left corner, regardless of our axis direction.
+    // originOffset gives us the delta from the real origin to the origin in the axis direction.
+    final Offset mainAxisUnit, crossAxisUnit, originOffset;
+    final bool addExtent;
+    switch (applyGrowthDirectionToAxisDirection(constraints.axisDirection, constraints.growthDirection)) {
+      case AxisDirection.up:
+        mainAxisUnit = const Offset(0.0, -1.0);
+        crossAxisUnit = const Offset(1.0, 0.0);
+        originOffset = offset + Offset(0.0, geometry!.paintExtent);
+        addExtent = true;
+      case AxisDirection.right:
+        mainAxisUnit = const Offset(1.0, 0.0);
+        crossAxisUnit = const Offset(0.0, 1.0);
+        originOffset = offset;
+        addExtent = false;
+      case AxisDirection.down:
+        mainAxisUnit = const Offset(0.0, 1.0);
+        crossAxisUnit = const Offset(1.0, 0.0);
+        originOffset = offset;
+        addExtent = false;
+      case AxisDirection.left:
+        mainAxisUnit = const Offset(-1.0, 0.0);
+        crossAxisUnit = const Offset(0.0, 1.0);
+        originOffset = offset + Offset(geometry!.paintExtent, 0.0);
+        addExtent = true;
+    }
 
+    RenderSliver? child = lastChild;
     while (child != null) {
+      final double mainAxisDelta = childMainAxisPosition(child);
+      final double crossAxisDelta = childCrossAxisPosition(child);
+      Offset childOffset = Offset(
+        originOffset.dx + mainAxisUnit.dx * mainAxisDelta + crossAxisUnit.dx * crossAxisDelta,
+        originOffset.dy + mainAxisUnit.dy * mainAxisDelta + crossAxisUnit.dy * crossAxisDelta,
+      );
+      if (addExtent) {
+        childOffset += mainAxisUnit * child.geometry!.paintExtent;
+      }
+
       if (child.geometry!.visible) {
-        final SliverPhysicalParentData childParentData = child.parentData! as SliverPhysicalParentData;
-        context.paintChild(child, offset + childParentData.paintOffset);
+        context.paintChild(child, childOffset);
       }
       child = childBefore(child);
```

Comment:
I don't completely understand this. Are we cycling backwards through every sliver group child, even though at some point they will no longer be `child.geometry!.visible`? It seems like a short-circuit would be useful although maybe no short-circuit is possible?

Reply:
We paint them in reverse because the group supports overlapping widgets like pinned app bars. Paint order stacks, so if we painted from first to last, a pinned app bar would be covered by the slivers it is meant to overlap. We only paints slivers that are visible to be efficient.

Reply:
I think what you're saying is that a short-circuit exit from the loop isn't possible because the visibility of one child doesn't imply anything about previous children?

Reply:
Correct. :)